cloud_blobstore/__init__.py

```diff
@@ -1,6 +1,74 @@
 import typing
 
 
+class PagedIter:
+    """
+    Provide an iterator that will iterate over every object, filtered by
+    prefix and delimiter. Alternately continue iteration with token and
+    key (marker).
+    """
+
+    def __init__(
+            self,
+            prefix: str=None,
+            delimiter: str=None,
+            marker: str=None,
+            token: str=None,
+            k_page_max: int=None
+    ):
+        self.kwargs = dict()
+        self.token = None
+        self.marker = None
+
+    def get_api_response(self, next_token):
+        """
+        Make blobstore-specific list api request.
+        """
+        raise NotImplementedError()
+
+    def get_listing_from_response(self, resp):
+        """
+        Retrieve blob key listing from blobstore response.
+        """
+        raise NotImplementedError()
+
+    def get_next_token_from_response(self, resp):
+        """
+        Retrieve opaque continuation token from blobstore response.
+        """
+        raise NotImplementedError()
+
+    def __iter__(self):
+        """
+        Iterate over the blobs, saving page tokens and blob key markers as
+        needed in order to continue listing where one left off.
+        """
+        next_token = self.token
+
+        while True:
+            self.token = next_token
+            resp = self.get_api_response(next_token)
+            listing = self.get_listing_from_response(resp)
+            next_token = self.get_next_token_from_response(resp)
+
+            i = 0
+            if self.marker:
+                try:
+                    i = 1 + next(i for (i, key) in enumerate(listing) if key == self.marker)
+                    listing = listing[i:]
+                except StopIteration:
+                    raise BlobPagingError('Marker not found in this page')
+
+            for key in listing:
+                self.marker = key
+                yield self.marker
+            else:
```

Comment:
is this else necessary?

Reply:
Yes, otherwise marker will be present for the next page of results, if there is one.

Reply:
`for..else` without `break` with no empty list special case is meaningless. I think you should just remove the `else:` and dedent the line that is in the `else:` block.

Reply:
Ah, I see.

Reply:
Actually.. This is a generator with a yield statement.

Marker is being set at every iteration. If the generator is broken (probably by a consumer of this generator), marker will be `not None`, and can be read by the user.

If the generator finishes, `self.marker` should be set to `None` so it does not interfere with the next page of results.

Reply:
I'm not sure what you mean by the generator being broken.

If the consumer of the generator does not finish consuming the output, then marker should _not_ be `None`.  If the consumer does finish consuming the output, then it does not matter.

```
In [10]: def foo_no_else():
    ...:     for x in range(5):
    ...:         yield x
    ...:     print("GOT TO THE END OF THE LOOP NO ELSE")
    ...:     

In [11]: def foo_else():
    ...:     for x in range(5):
    ...:         yield x
    ...:     else:
    ...:         print("GOT TO THE END OF THE LOOP")
    ...: 

In [12]: for z in foo_no_else():
    ...:     print(z)
    ...: 
0
1
2
3
4
GOT TO THE END OF THE LOOP NO ELSE

In [13]: for z in foo_else():
    ...:     print(z)
    ...: 
0
1
2
3
4
GOT TO THE END OF THE LOOP

In [14]:

Reply:
I mean the iteration is not completed.

There is an outer loop to consider:

    In [8]: def foo_no_else():
       ...:     expectations = 'some'
       ...:     for i in range(2):
       ...:         yield i
       ...:         for _ in range(3):
       ...:             pass
       ...:     print(expectations)
       ...:     

    In [9]: def foo_else():
       ...:     expectations = 'some'
       ...:     for i in range(2):
       ...:         yield i
       ...:         for _ in range(3):
       ...:             pass
       ...:         else:
       ...:             expectations = None
       ...:     print(expectations)
       ...:     
       ...:     

    In [10]: for z in foo_no_else():
        ...:     print(z)
        ...:     
    0
    1
    some
    
    In [11]: for z in foo_else():
        ...:     print(z)
        ...:     
    0
    1
    None

Reply:
@xbrianh I'm not sure what you mean either. The point I was trying to make is not affected by generator semantics. An `else` statement in a `for..else` loop will always run if the loop does not contain a `break` statement, so you might as well remove the `else` and dedent the block. It's impossible for the caller to affect the generator's control flow statements (i.e., introduce a "break") - the caller and the generator are in different, isolated frames.

You can verify this by removing the else: and dedenting the following line (`expectations = None`) one level in your example.

Reply:
Ok, I understand my confusion.